src/librustc_typeck/collect.rs

```diff
@@ -740,11 +740,14 @@ fn trait_def<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     }
 
     let def_path_hash = tcx.def_path_hash(def_id);
-    let has_default_impl = tcx.hir.trait_is_auto(def_id);
+    let is_auto = match item.node {
+        hir::ItemTrait(hir::IsAuto::Yes, ..) => true,
+        _ => tcx.hir.trait_is_auto(def_id),
```

Comment:
Once we make `hir::IsAuto::Yes` be the canonical way to encode hir traits, this code can be simplified, as can quite a bit of other code I imagine.